ethcore/src/state/mod.rs

```diff
@@ -539,8 +542,55 @@ impl<B: Backend> State<B> {
 			|a| a.as_ref().and_then(|account| account.storage_root().cloned()))
 	}
 
-	/// Mutate storage of account `address` so that it is `value` for `key`.
-	pub fn storage_at(&self, address: &Address, key: &H256) -> TrieResult<H256> {
+	/// Get the value of storage at a specific checkpoint.
+	pub fn checkpoint_storage_at(&self, checkpoint_index: usize, address: &Address, key: &H256) -> TrieResult<Option<H256>> {
+		let checkpoints = self.checkpoints.borrow();
+		let checkpoints_len = checkpoints.len();
+		if let Some(ref checkpoint) = checkpoints.get(checkpoint_index) {
+			if let Some(entry) = checkpoint.get(address) {
+				match entry {
+					Some(entry) => {
+						match entry.account {
+							Some(ref account) => {
+								if let Some(value) = account.cached_storage_at(key) {
+									Ok(Some(value))
+								} else {
+									// This key has checkpoint entry, but not in the entry's cache.
+									// So it's not modified at all.
+									Ok(Some(self.original_storage_at(address, key)?))
+								}
+							},
+							None => {
+								// The account didn't exist at that point. Return empty value.
+								Ok(Some(H256::new()))
+							},
+						}
+					},
+					None => {
+						// The value was not cached at that checkpoint, meaning it was not modified.
```

Comment:
Perhaps it was modified in one of the previous checkpoints?

Reply:
If we find `None` as existing for a checkpoint entry, then it means that previously when swapped from cache, the cache doesn't have that value. This means that it's not modified by this `State<B>` at all.

Reply:
Hmm that was L563. Let me check this.